### DISCUSSION TITLE
Meeting Minutes for October 3, 2024

### DISCUSSION THREAD

**Discussion by dezidizon** [CVWRSM/cvwrsm#139] [Meeting Minutes]
**Agenda:**
- Beginning work on IWFM and IDC Migration.
- Updates for the WRIMS OSGI Bundle/Plugin Updates.
    - Technical walkthrough of the CI/CD work for building the plugins for wrims, gov.ca.dwr.hecdssvue, and the 3rd party OSGI bundle
    - Identify DWR resource to review pull requests 
- Follow-up on coordination for the CBC solver work.
- Identify DWR resource who knows the technical parts for the HEC-DSS Excel Addin CI/CD.
- Follow-up on reported EPPT issues. 

**Open Action Items:**
- Discuss the testing utility for the IWFM and IDC automatic build.
- Coordinate meeting with @hxiedwr to cherry-pick commits made in the master branch into the wrims-devops branch.
- Discuss maintaining a branch with WRIMS code using HecLib 6 and setting the wrims-devops branch into the new default branch.
- Prepare level of effort, budget, and schedule to perform work for the following project(s):
    - WRIMS 3 Development 

**Discussions:**

- Cherry-picking commits from the [master branch](https://github.com/CentralValleyModeling/wrims/tree/master) into the [wrims-devops branch](https://github.com/CentralValleyModeling/wrims/tree/Feature/wrims-devops):
    - @hxiedwr requested an estimate on when the commits from the master branch will be cherry-picked into the wrims-devops branch.
    - The commits in the master branch will be cherry-picked into wrims-devops after completing the work on the WRIMS OSGI Bundle/Plugin updates.
    - The cherry-picking process will be coordinated closely with @hxiedwr in a meeting. 
- Maintaining a branch for WRIMS code using HecLib 6 and setting the wrims-devops branch as the new default branch:
    - @hxiedwr discussed maintaining a branch with WRIMS code using HecLib6 and setting the wrims-devops branch as the new default branch. The intention is to maintain separate branches for WRIMS code using HecLib 6 and WRIMS code using HecLib 7. 
    - @dwr-zroy recommended creating a tag for the head revision of the master branch so that the branch can be archived. A new branch can be created from that tag if development needs to occur specific to WRIMS using HecLib 6.
    - @rma-psmorris recommended renaming the "master" branch to the "main" branch to match naming convention used by USACE.
- IWFM and IDC Migration:
    - @stephen-rma will begin work on the IWFM and IDC Migration.
    - @ecdogrul will provide the latest version of the SVN repositories for @stephen-rma to work with. After the migration work is done, a newer version of the repositories will be shared again so that updates can be integrated into the Git repositories.
    - @stephen-rma will meet with @ecdogrul when there are representative converted repositories done to review and demonstrate how these would be expected to be seen in GitHub.
- IWFM and IDC Testing Utility Discussion:
    - @ecdogrul discussed that part of setting up the automated build system for IWFM and IDC should include an automated testing utility.
    - This was not included in the work plan submitted to @ecdogrul as the RMA team wanted to focus on starting the IWFM and IDC migration from SVN to GitHub.
    - This will be discussed once the migration work has been completed.
- WRIMS OSGI Bundles/Plugins Updates:
    - With work on preparing the WRIMS OSGI Bundles and Plugins in progress, @rma-psmorris asked how the DWR staff can be engaged with the CI/CD workflow.
        - @hxiedwr will be performing the integration testing of the artifacts that RMA is producing through the CI/CD process.
        - Once the work is done with the WRIMS plugin updates, RMA will work with DWR staff to go over the DevOps process. There are known issues with WRIMS that can be used walkthrough the DevOps process 
- CBC Solver:
    - The DWR team is working on getting the CBC solver workflow into a good state. Then @dwr-ycheng will coordinate a meeting with RMA to discuss this work.
- HEC-DSS Excel Addin:
    - @dwr-ycheng will follow-up with the DWR resources to join the discussion for the HEC-DSS Excel Addin CI/CD work.
- EPPT:
    - RMA will follow-up with @inazrul regarding the EPPT issues that were reported.

**Closed Action Items:**
- @dwr-ycheng  to follow-up on whether RMA should move ahead with preparing the installer for the HEC-DSS Excel Addin.
- @rma-psmorris to organize team to help with the CBC Solver.
- RMA to follow-up with DWR to discuss work for the CBC Solver.
- Prepare level of effort, budget, and schedule to perform work for the following project(s):
    - IWFM and IDC Automated Build System and Source Control

**Comment by hxiedwr** [comment]
Creating a new branch duplicated based on master (heclib 6) and then merging wrims_devops (heclib 7) to master proposed by RMA staffs might be a better approach. Or we can do tagging if it works.

A "feature/solver" branch was developed based on the "master" branch (heclib 6). With the proposed methods above, it will be easier to merge it to the new master (heclib 7) than merging master to wrims_devops branch.
________________________________
From: dezidizon ***@***.***>
Sent: Thursday, October 3, 2024 5:27:41 PM
To: CentralValleyModeling/wrims ***@***.***>
Cc: Xie, ***@***.*** ***@***.***>; Mention ***@***.***>
Subject: [CentralValleyModeling/wrims] Meeting Minutes for October 3, 2024 (Discussion #139)


Agenda:

  *   Beginning work on IWFM and IDC Migration.
  *   Updates for the WRIMS OSGI Bundle/Plugin Updates.
     *   Technical walkthrough of the CI/CD work for building the plugins for wrims, gov.ca.dwr.hecdssvue, and the 3rd party OSGI bundle
     *   Identify DWR resource to review pull requests
  *   Follow-up on coordination for the CBC solver work.
  *   Identify DWR resource who knows the technical parts for the HEC-DSS Excel Addin CI/CD.
  *   Follow-up on reported EPPT issues.

Open Action Items:

  *   Discuss the testing utility for the IWFM and IDC automatic build.
  *   Coordinate meeting with @hxiedwr<https://github.com/hxiedwr> to cherry-pick commits made in the master branch into the wrims-devops branch.
  *   Discuss maintaining a branch with WRIMS code using HecLib 6 and setting the wrims-devops branch into the new default branch.
  *   Prepare level of effort, budget, and schedule to perform work for the following project(s):
     *   WRIMS 3 Development

Discussions:

  *   @hxiedwr<https://github.com/hxiedwr> requested an estimate on when the commits from the master branch will be cherry-picked into the wrims-devops branch.
     *   The commits in the master branch will be cherry-picked into wrims-devops after completing the work on the WRIMS OSGI Bundle/Plugin updates.
     *   The cherry-picking process will be coordinated closely with @hxiedwr<https://github.com/hxiedwr> in a meeting.
  *   @hxiedwr<https://github.com/hxiedwr> discussed maintaining a branch with WRIMS code using HecLib6 and setting the wrims-devops branch as the new default branch.
     *   @dwr-zroy<https://github.com/dwr-zroy> recommended creating a tag for the head revision of the master branch so that the branch can be archived. A new branch can be created from that tag if development needs to occur specific to WRIMS using HecLib 6.
  *   @stephen-rma<https://github.com/stephen-rma> will begin work on the IWFM and IDC Migration.
     *   @ecdogrul<https://github.com/ecdogrul> will provide the latest version of the SVN repositories for @stephen-rma<https://github.com/stephen-rma> to work with. After the migration work is done, a newer version of the repositories will be shared again so that updates can be integrated into the Git repositories.
     *   @stephen-rma<https://github.com/stephen-rma> will meet with @ecdogrul<https://github.com/ecdogrul> when there are representative converted repositories done to review and demonstrate how these would be expected to be seen in GitHub.
  *   @ecdogrul<https://github.com/ecdogrul> discussed that part of setting up the automated build system for IWFM and IDC should include an automated testing utility.
     *   This was not included in the work plan submitted to @ecdogrul<https://github.com/ecdogrul> as the RMA team wanted to focus on starting the IWFM and IDC migration from SVN to GitHub.
     *   This will be discussed once the migration work has been completed.
  *   With work on preparing the WRIMS OSGI Bundles and Plugins in progress, @rma-psmorris<https://github.com/rma-psmorris> asked how the DWR staff can be engaged with the CI/CD workflow.
     *   @hxiedwr<https://github.com/hxiedwr> will be performing the integration testing of the artifacts that RMA is producing through the CI/CD process.
     *   Once the work is done with the WRIMS plugin updates, RMA will work with DWR staff to go over the DevOps process. There are known issues with WRIMS that can be used walkthrough the DevOps process
  *   The DWR team is working on getting the CBC solver workflow into a good state. Then @dwr-ycheng<https://github.com/dwr-ycheng> will coordinate a meeting with RMA to discuss this work.
  *   @dwr-ycheng<https://github.com/dwr-ycheng> will follow-up with the DWR resources to join the discussion for the HEC-DSS Excel Addin CI/CD work.
  *   RMA will follow-up with @inazrul<https://github.com/inazrul> regarding the EPPT issues that were reported.

Closed Action Items:

  *   @dwr-ycheng<https://github.com/dwr-ycheng> to follow-up on whether RMA should move ahead with preparing the installer for the HEC-DSS Excel Addin.
  *   @rma-psmorris<https://github.com/rma-psmorris> to organize team to help with the CBC Solver.
  *   RMA to follow-up with DWR to discuss work for the CBC Solver.
  *   Prepare level of effort, budget, and schedule to perform work for the following project(s):
     *   IWFM and IDC Automated Build System and Source Control

**Reply by dwr-zroy** [reply]
Because I think visually, here is what I was imagining when I brought up the 'tagging' idea.

## Approximate current state of the repo

We have branch `master` which is default. It lags behind `devops` in some ways, but there are a few cherry picks we will need to do.
```mermaid
%%{init: {'themeVariables': {'git0': '#377eb8', 'git1': '#4daf4a'}, 'gitGraph': {'mainBranchName': 'master'}} }%%
gitGraph
commit
branch devops
commit
checkout master
commit id: "commit to cherrypick"
commit id: "latest DSS 6"
checkout devops
commit id: "CI/CD complete"
```

## After tagging the master branch, cherry-picking, and changing default
We first cherry pick the changes from `master` onto `devops`, and then tag the current commit of `master` and delete the branch, Note that the commit history isn't lost, because the tag still points to that latest commit. We then change the default branch in the git configuration.
```mermaid
%%{init: { 'themeVariables': {'git1': '#377eb8', 'git0': '#4daf4a'},'gitGraph': {'mainBranchName': '_', 'mainBranchOrder': 2}} }%%
gitGraph
commit
branch devops order: 1
commit
checkout _
commit id: "commit to cherrypick"
commit id: "latest DSS 6" tag: "archive-dss-6"
checkout devops
commit id: "CI/CD complete"
checkout devops
cherry-pick id:"commit to cherrypick"
```


## Future development occurs on `devops`
As we move forward, additional work is just done on `devops` (via pull requests from other branches ideally).
```mermaid
%%{init: {'themeVariables': {'git1': '#377eb8', 'git0': '#4daf4a'}, 'gitGraph': {'mainBranchName': '_', 'mainBranchOrder': 2}} }%%
gitGraph
commit
branch devops order: 1
commit
checkout _
commit id: "commit to cherrypick"
commit id: "latest DSS 6" tag: "archive-dss-6"
checkout devops
commit id: "CI/CD complete"
checkout devops
cherry-pick id:"commit to cherrypick"
commit id: "feature-1"
commit id: "bugfix-1"
commit id: "etc"
```

## Ability to open up the archived DSS-6 branch again if needed
If we need to support the DSS-6 version of WRIMS though, we can still go back to the tag, create a new branch, and move it forward as needed. After that work is done we would probably want to tag/delete branch again.
We can do that by checking out the tag, creating a new branch there, and moving forward.

```mermaid
%%{init: { 'themeVariables': {'git1': '#377eb8', 'git0': '#4daf4a'},'gitGraph': {'mainBranchName': 'archived-dss-6', 'mainBranchOrder': 2}} }%%
gitGraph
commit
branch devops order: 1
commit
checkout archived-dss-6
commit id: "commit to cherrypick"
commit id: "latest DSS 6" tag: "archive-dss-6"
checkout devops
commit id: "CI/CD complete"
checkout devops
cherry-pick id:"commit to cherrypick"
commit id: "feature-1"
commit id: "bugfix-1"
commit id: "etc"
checkout archived-dss-6
commit id: "support archived WRIMS"
```

**Comment by hxiedwr** [comment]
With committing changes from master to devops, it may be hard for "feature/solver" branch to be merged to "devops" branch later on.
"feature/solver" branch occurred before cherrypick commit from the "master" branch. It need to be merged to "devops" branch some time after the cherrypick commit.

**Reply by dwr-zroy** [reply]
In that case, we could do something like this:

## Plan including `feature/solver`
Cherry pick from `master` to `feature/solver` before the merge.

```mermaid
%%{init: { 'themeVariables': {'git1': '#377eb8', 'git0': '#4daf4a'},'gitGraph': {'mainBranchName': 'master', 'mainBranchOrder': 2}} }%%
gitGraph
commit
branch solver order: 3
commit
checkout master
commit

branch devops order: 1
commit

checkout master
commit id: "commit to cherrypick"
commit id: "latest DSS 6" tag: "archive-dss-6"

checkout devops
commit id: "CI/CD complete"

checkout devops
cherry-pick id:"commit to cherrypick"

checkout solver
cherry-pick id:"commit to cherrypick"
commit id: "resolve any issues"

checkout devops
merge solver id: "merge solver after cherry-picking"
commit id: "feature-1"
commit id: "bugfix-1"
commit id: "etc"
```


I'm definitely ignorant of the technical complexity of the cherry-pick and the merge process. So if that is the prohibiting factor, that's good to know.
